Documentation/proposals/202310-automated-sharding.md

```diff
@@ -0,0 +1,130 @@
+
+# How
+
+Today, there are a few strategies to measure load on top of Prometheus instances.
+
+1. Perhaps you can use [Kubernetes Metrics API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/#metrics-api) to measure load in terms of CPU and/or Memory usage.
+
+2. Another strategy might be measure amount of samples ingested per second
+
+- Prometheus Agents: `rate(prometheus_agent_samples_appended_total[5m])`
+- Prometheus Servers: `rate(prometheus_tsdb_head_samples_appended_total[5m])`
+
+3. Or even measuring samples queried per second for Prometheus instances under heavy query load.
+
+- Prometheus Servers: `rate(prometheus_engine_query_samples_total[5m])`
+
+Modern Custom Horizontal Pod Autoscalers (HPAs), such as Keda are already capable of [scaling CRDs](https://keda.sh/docs/2.11/concepts/scaling-deployments/#scaling-of-custom-resources) based on different parameters, such as [CPU, Memory](https://keda.sh/docs/2.11/scalers/cpu/) or even [custom Prometheus metrics](https://keda.sh/docs/2.11/scalers/prometheus/).
+
+### Scale Status subresource
+
+When working with CRDs, common HPAs (e.g. Keda) depends on a Status subresource called [Scale](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource). Unfortunately, Prometheus-Operator resources still don't implement this field. There was an attempt to implement this in https://github.com/prometheus-operator/prometheus-operator/pull/4735, but was reverted because the PR was scaling replicas instead of shards. Although scaling replicas can help with High-Availability, it is expensive and hard to manage since it duplicates scrapes while not reducing the load on top of Prometheus because all replicas still scrape the same targets. Sharding serves as a better autoscaling strategy since the load is spread into all instances instead of getting duplicated.
+
+With only this change, Prometheus Agents can already be horizontally scaled without problems, but for Prometheus Servers it gets a little more complicated.
+
+### Graceful scale down of Prometheus Servers
+
+Prometheus Servers don't behave like the usual stateless applications. A few cases require Prometheus pods to still be available even after a scale down event is triggered:
+
+* Thanos proxying queries back to the sidecars to retrieve data that still wasn't uploaded to object storage.
+* Prometheus servers with long retention periods still hold data that users might want to query.
+
+If we simply delete Prometheus pods during scale-down, data loss will be a common problem, so a more complex scale-down strategy needs to be used here.
+
+We propose that Prometheus-Operator introduces a new field to the Prometheus Server CRDs:
+
+```yaml
+apiVersion: monitoring.coreos.com
+kind: Prometheus
+metadata:
+  name: example
+spec:
+  autoscaling:
+    scaleDownStrategy: force|preventDataLoss # Default: force
```

Comment:
I'd suggest `graceful` instead of `preventDataLoss` and I'm bit torn between `immediate` over `force`.

wdyt?

Reply:
Agreed, sounds better!